source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,591 @@
+
+      // The list of boundary indicators for which we need to set
+      // zero mesh velocities, which means the zero/prescribed Stokes velocity
+      // boundaries minus those that are listed as Additional
+      // tangential boundary indicators.
+
+      // The list of prescribed Stokes velocity boundary indicators.
+      const std::map<types::boundary_id, std::pair<std::string,std::vector<std::string> > > active_boundary_velocity_indicators =
+        this->get_boundary_velocity_manager().get_active_boundary_velocity_names();
+      std::set<types::boundary_id> prescribed_boundary_velocity_indicators;
+      //for (std::map<types::boundary_id, std::pair<std::string, std::vector<std::string> > >::const_iterator p = active_boundary_velocity_indicators.begin();
+      //     p != active_boundary_velocity_indicators.end(); ++p)
+      for (auto p = active_boundary_velocity_indicators.begin();
+           p != active_boundary_velocity_indicators.end(); ++p)
+        prescribed_boundary_velocity_indicators.insert(p->first);
+      // The list of zero Stokes velocity boundary indicators.
+      std::set<types::boundary_id> x_zero_boundary_indicators = this->get_boundary_velocity_manager().get_zero_boundary_velocity_indicators();
+      x_zero_boundary_indicators.insert(prescribed_boundary_velocity_indicators.begin(), prescribed_boundary_velocity_indicators.end());
+      // Remove Additional tangential boundary indicators from the list of prescribed velocity
+      // boundary indicators.
+      for (std::set<types::boundary_id>::const_iterator p = x_zero_boundary_indicators.begin();
+           p != x_zero_boundary_indicators.end(); ++p)
+        if (boundary_ids.find(*p) == boundary_ids.end())
+          if (additional_tangential_mesh_boundary_indicators.find(*p) == additional_tangential_mesh_boundary_indicators.end())
+            {
+              VectorTools::interpolate_boundary_values (mesh_deformation_dof_handler, *p,
+                                                        dealii::Functions::ZeroFunction<dim>(dim), matrix_constraints);
+            }
+
+      // The list of boundary indicators for which we need to set
+      // no_normal_flux_constraints, which means all
+      // minus the diffusion mesh deformation boundary indicators
+      // and minus the zero/prescribed Stokes velocity boundary indicators.
+      std::set<types::boundary_id> x_no_flux_boundary_indicators = this->get_boundary_velocity_manager().get_tangential_boundary_velocity_indicators();
+      x_no_flux_boundary_indicators.insert(additional_tangential_mesh_boundary_indicators.begin(),additional_tangential_mesh_boundary_indicators.end());
+      for (std::set<types::boundary_id>::const_iterator p = x_no_flux_boundary_indicators.begin();
+           p != x_no_flux_boundary_indicators.end(); ++p)
+        if (boundary_ids.find(*p) != boundary_ids.end())
+          {
+            x_no_flux_boundary_indicators.erase(*p);
```

Comment:
I think you can do this more easily with `std::set_minus`. If you grep for it in the code base, @gassmoeller used this somewhere recently.